File: VBNET/IODevices/IOmsg.vb
Option Strict Off
Option Explicit On
Imports System.Drawing
Imports System.Windows.Forms
Imports IODevices
Namespace IODeviceForms  'internal namespace

    Friend Class IOmsgForm  'one form per device
        Inherits System.Windows.Forms.Form

        Private query As IOQuery  'set by calling thread
        Public Sub showit(ByVal q As IOQuery) 'called by gpib thread

            Dim cberr As Boolean = (q.status And 256) <> 0
            If cberr Then q.status -= 256

            query = q
            Text = q.device.devname + " error"
            msg1.ForeColor = Color.Red
            lbl_retry.ForeColor = Color.Red

            lbl_retry.Visible = True
            'cmd_abort.Visible = False

            If query.status > 0 Then
                If (query.status And 2) = 0 Then             'bit 2 on send(0)/recv(1)
                    msg1.Text = "error while sending data to " + query.device.devname
                Else
                    msg1.Text = "error while receiving data from " + query.device.devname
                End If
                'compose info:

                txt.Text = ""


                txt.Text &= "address: " & query.device.devaddr
                txt.Text &= vbCrLf & "command: " & query.cmd

                If (query.status And 4) > 0 Then
                    txt.Text &= vbCrLf & "interface returned error n°" & query.errcode.ToString
                End If

                txt.Text &= vbCrLf & query.errmsg
                'if retry:
                If query.task.retry Then
                    lbl_retry.Text = " retrying ..."
                    cmd_abort.Visible = True
                    Timer1.Enabled = True
                Else
                    lbl_retry.Text = "IO operation abandoned"
                    cmd_abort.Visible = False

                    Timer1.Enabled = False
                End If

            ElseIf cberr Then
                msg1.Text = query.device.devname + " : error in callback "
                txt.Text = "unhandled exception in user callback function"

                txt.Text &= vbCrLf & "command: " & query.cmd

                txt.Text &= vbCrLf & query.errmsg

                lbl_retry.Text = "IO operation completed"
                cmd_abort.Visible = False

                Timer1.Enabled = False

            End If





            Show()

        End Sub

        Private Sub cmd_abort_Click(ByVal eventSender As System.Object, ByVal eventArgs As System.EventArgs) Handles cmd_abort.Click

            query.AbortRetry()

            Close()

        End Sub



        Private Sub cmd_ok_Click(ByVal eventSender As System.Object, ByVal eventArgs As System.EventArgs) Handles cmd_ok.Click

            Close()

        End Sub


        Private Sub Timer1_Tick(ByVal eventSender As System.Object, ByVal eventArgs As System.EventArgs) Handles Timer1.Tick
            'make it blinking

            If lbl_retry.Visible Then
                lbl_retry.Visible = False
            Else
                lbl_retry.Visible = True
            End If

            Me.Show()

        End Sub

        Private Sub cmd_clear_Click(ByVal sender As System.Object, ByVal e As System.EventArgs)

        End Sub

        Private Sub cmd_abortall_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles cmd_abortall.Click

            query.AbortRetry()  'abort this (blocking or async)

            query.AbortAll() 'abort all async commands in queue
            Close()
            'Hide()
        End Sub

        Private Sub IOmsgForm_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load
            txt.Multiline = True
        End Sub


        Public Sub shutdownmsg(ByVal devname As String) 'called by iodevice when shutting down

            Try
                Text = devname
                msg1.ForeColor = Color.Red
                lbl_retry.Visible = False

                cmd_abort.Visible = False
                cmd_abortall.Visible = False
                cmd_ok.Visible = False

                msg1.Text = "shutting down " + devname + " ..."


                Show()
                WindowState = FormWindowState.Normal
                BringToFront()

            Catch ex As Exception

            End Try



        End Sub

        Private Sub txt_TextChanged(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles txt.TextChanged

        End Sub
    End Class
End Namespace

File: C#/IODevices/IOmsg.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Windows.Forms;

using IODevices;


namespace IODeviceForms
{
    internal partial class IOmsgForm : Form
    {
        public IOmsgForm()
        {
            InitializeComponent();
        }
    


    
			//set by calling thread
		private IOQuery query;

        private string CrLf =System.Text.Encoding.UTF8.GetString(new byte[2] { 13, 10 }); //CrLf;
		//called by gpib thread
		public void showit(IOQuery q)
		{

			bool cberr = (q.status & 256)!=0;
			if (cberr)
				q.status -= 256;

			query = q;
			Text = q.device.devname + " error";
			msg1.ForeColor = Color.Red;
			lbl_retry.ForeColor = Color.Red;

			lbl_retry.Visible = true;
			//cmd_abort.Visible = False

			if (query.status > 0) {
				//bit 2 on send(0)/recv(1)
				if ((query.status & 2) == 0) {
					msg1.Text = "error while sending data to " + query.device.devname;
				} else {
					msg1.Text = "error while receiving data from " + query.device.devname;
				}
				//compose info:

				txt.Text = "";


				txt.Text += "address: " + query.device.devaddr;
				txt.Text += CrLf + "command: " + query.cmd;

                if ((query.status & 4) > 0)
                { txt.Text += CrLf + "interface returned error n°" + query.errcode.ToString(); }
				
                txt.Text += CrLf + query.errmsg;
				//if retry:
				if (query.task.retry) {
					lbl_retry.Text = " retrying ...";
					cmd_abort.Visible = true;
					Timer1.Enabled = true;
				} else {
					lbl_retry.Text = "IO operation abandoned";
					cmd_abort.Visible = false;

					Timer1.Enabled = false;
				}

			} else if (cberr) {
				msg1.Text = query.device.devname + " : error in callback ";
				txt.Text = "unhandled exception in user callback function";

				txt.Text += CrLf + "command: " + query.cmd;

				txt.Text += CrLf + query.errmsg;

				lbl_retry.Text = "IO operation completed";
				cmd_abort.Visible = false;

				Timer1.Enabled = false;

			}





			Show();

		}


		private void cmd_abort_Click(System.Object eventSender, System.EventArgs eventArgs)
		{
			query.AbortRetry();

			Close();

		}




		private void cmd_ok_Click(System.Object eventSender, System.EventArgs eventArgs)
		{
			Close();

		}


		private void Timer1_Tick(System.Object eventSender, System.EventArgs eventArgs)
		{
			//make it blinking

			if (lbl_retry.Visible) {
				lbl_retry.Visible = false;
			} else {
				lbl_retry.Visible = true;
			}

			this.Show();

		}


		private void cmd_clear_Click(System.Object sender, System.EventArgs e)
		{
		}


		private void cmd_abortall_Click(System.Object sender, System.EventArgs e)
		{
			query.AbortRetry();
			//abort this (blocking or async)

			query.AbortAll();
			//abort all async commands in queue
			Close();
			//Hide()
		}

		private void IOmsgForm_Load(System.Object sender, System.EventArgs e)
		{
			txt.Multiline = true;
		}


		//called by iodevice when shutting down
		public void shutdownmsg(string devname)
		{

			try {
				Text = devname;
				msg1.ForeColor = Color.Red;
				lbl_retry.Visible = false;

				cmd_abort.Visible = false;
				cmd_abortall.Visible = false;
				cmd_ok.Visible = false;

				msg1.Text = "shutting down " + devname + " ...";


				Show();
				WindowState = FormWindowState.Normal;
				BringToFront();


			} catch (Exception ) {
			}



        }

        private void txt_TextChanged(object sender, EventArgs e)
        {

        }

	}

}

